src/Data/Text/Markdown.hs

```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Data.Text.Markdown where
+
+import CMarkGFM qualified as CMark
+import Control.Applicative ((<|>))
+import Control.Monad (void)
+import Data.Function ((&))
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Yaml
+import Swarm.Language.Module (moduleAST)
+import Swarm.Language.Parse (readTerm)
+import Swarm.Language.Pipeline (ProcessedTerm (..), prettyTypeErr, processParsedTerm)
+import Swarm.Language.Pretty (prettyText)
+import Swarm.Language.Syntax (Syntax)
+
+data Node
+  = Node NodeType [Node]
+  | LeafText Text
+  | LeafCode Syntax
+  | LeafCodeBlock String Syntax
+  deriving (Eq, Show)
+
+data NodeType
+  = Document
+  | Paragraph
+  deriving (Eq, Show)
+
+instance ToJSON Node where
+  toJSON = String . toText
+
+instance FromJSON Node where
+  parseJSON v =
+    withArray "markdown paragraphs" (foldMap parseJSON) v
+      <|> withText "markdown text" fromTextM v
+instance Semigroup Node where
+  Node Document cs1 <> Node Document cs2 = Node Document (cs1 <> cs2)
+  Node Document cs1 <> n2 = Node Document (cs1 <> [n2])
+  n1 <> Node Document cs2 = Node Document (n1 : cs2)
+  n1 <> n2 = Node Document [n1, n2]
+
+instance Monoid Node where
+  mempty = Node Document []
+
+fromText :: Text -> Node
+fromText = either error id . fromTextE
+
+fromTextM :: MonadFail m => Text -> m Node
+fromTextM = either fail pure . fromTextE
+
+fromTextE :: Text -> Either String Node
+fromTextE = fromGeneralNode . CMark.commonmarkToNode [CMark.optSourcePos] []
+
+-- TODO: use transformer kind of Error? (report all errors)
+fromGeneralNode :: CMark.Node -> Either String Node
+fromGeneralNode (CMark.Node p t cs) = case t of
+  CMark.DOCUMENT -> Node Document <$> mapM fromGeneralNode cs
+  CMark.PARAGRAPH -> Node Paragraph <$> mapM fromGeneralNode cs
+  CMark.SOFTBREAK -> leaf $ LeafText "\n" -- ?
```

Comment:
According to https://github.github.com/gfm/#soft-line-breaks it seems like a `SOFTBREAK` is what we get at the end of a physical line of parsed text which does not end with two or more spaces.  Typically we ignore such breaks and just reflow the text, so I would think we should not generate a `LeafText "\n"` here, unless I am misunderstanding something.

Reply:
I think we might need to use a space instead of newline here.

Reply:
Note to self: check that spacing is preserved.